src/main/java/io/cdap/plugin/gcp/bigquery/relational/BigQueryRelation.java

```diff
@@ -1,117 +1,409 @@
+   */
+  public String getTransformExpression() {
+    return expressionSupplier.get();
+  }
+
+  /**
+   * Get columns defined in this relation
+   * @return Columns defined in this relation.
+   */
+  public Set<String> getColumns() {
+    return columns;
+  }
+
+  /**
+   * Sets input datasets for this instance and the parent instance (if defined)
+   */
+  public void setInputDatasets(Map<String, BigQuerySQLDataset> datasets) {
+    this.sourceDatasets = datasets;
+
+    // Propagate datasets into parent.
+    if (parent != null) {
+      parent.setInputDatasets(datasets);
+    }
+  }
+
+  /**
+   * Get the dataset name for this relation.
+   * @return dataset name
+   */
+  public String getDatasetName() {
+    return datasetName;
+  }
+
+  /**
+   * Get a new relation with a redefined dataset name.
+   * @param newDatasetName new dataset name for this relation.
+   * @return new Relation with the new dataset name.
+   */
+  public Relation setDatasetName(String newDatasetName) {
+    Map<String, Expression> selectedColumns = getSelectedColumns(columns);
+    // Build new transform expression and return new instance.
+    Supplier<String> supplier =
+      () -> buildNestedSelect(selectedColumns, getTransformExpression(), newDatasetName, null);
+    return new BigQueryRelation(newDatasetName, columns, this, supplier);
   }
 
   @Override
   public Relation setColumn(String column, Expression value) {
-    Map<String, Expression> newColumns = new HashMap<>(columns);
-    newColumns.put(column, value);
+    // check if expression is supported and valid
+    if (!supportsExpression(value)) {
+      return getInvalidRelation("Unsupported or invalid expression type.");
+    }
 
-    return new BigQueryRelation(datasetName, newColumns, filter, this);
+    Map<String, Expression> selectedColumns = getSelectedColumns(columns);
+    selectedColumns.put(column, value);
+
+    // Build new transform expression and return new instance.
+    Supplier<String> supplier =
+      () -> buildNestedSelect(selectedColumns, getTransformExpression(), datasetName, null);
+    return new BigQueryRelation(datasetName, selectedColumns.keySet(), this, supplier);
   }
 
   @Override
   public Relation dropColumn(String column) {
-    Map<String, Expression> newColumns = new HashMap<>(columns);
-    newColumns.remove(column);
+    // check if all expressions are supported and valid
+    if (!columns.contains(column)) {
+      return getInvalidRelation("Trying to remove non existing column in Relation: " + column);
+    }
+
+    // Remove column
+    Map<String, Expression> selectedColumns = getSelectedColumns(columns);
+    selectedColumns.remove(column);
 
-    return new BigQueryRelation(datasetName, newColumns, filter, this);
+    // Build new transform expression and return new instance.
+    Supplier<String> supplier =
+      () -> buildNestedSelect(selectedColumns, getTransformExpression(), datasetName, null);
```

Comment:
nit: you probably don't need to build a nested select here. You can simply drop a column from the list and use the original expression